src/TableFunctions/ITableFunctionDataLake.h

```diff
@@ -0,0 +1,177 @@
+#pragma once
+
+#include "config.h"
+
+#if USE_AWS_S3
+
+#include <TableFunctions/ITableFunction.h>
+#include <Storages/ExternalDataSourceConfiguration.h>
+
+#    include <filesystem>
+#    include <Access/Common/AccessFlags.h>
+#    include <Formats/FormatFactory.h>
+#    include <IO/S3Common.h>
+#    include <Interpreters/Context.h>
+#    include <Interpreters/evaluateConstantExpression.h>
+#    include <Interpreters/parseColumnsListForTableFunction.h>
+#    include <Parsers/ASTLiteral.h>
+#    include <Storages/StorageURL.h>
+#    include <Storages/checkAndGetLiteralArgument.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+}
+
+class Context;
+class TableFunctionS3Cluster;
+
+template <typename Name, typename Storage>
+class ITableFunctionDataLake : public ITableFunction
+{
+public:
+    static constexpr auto name = Name::name;
+    std::string getName() const override
+    {
+        return name;
+    }
+
+protected:
+    StoragePtr
+    executeImpl(const ASTPtr & /*ast_function*/, ContextPtr context, const std::string & table_name, ColumnsDescription /*cached_columns*/)
```

Comment:
let's move this code into cpp and add something like this https://github.com/ClickHouse/ClickHouse/blob/master/src/Common/NamedCollections/NamedCollections.cpp#L435 to resolve the templates.